cpp/include/types.hpp

```diff
@@ -0,0 +1,48 @@
+/*
+ * Copyright (c) 2018, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#ifndef TYPES_HPP
+#define TYPES_HPP
+
+#include <algorithm>
+#include <cassert>
+#include "cudf.h"
+
+namespace cudf {
+
+/**
+ * @brief A wrapper for a set of gdf_columns all with equal number of rows.
+ *
+ */
+struct table {
+  table(gdf_column* cols[], gdf_size_type num_cols)
+      : columns{cols}, num_columns{num_cols} {
+
+    assert(nullptr != cols[0]);
+
+    gdf_size_type const num_rows{cols[0]->size};
+
+    std::for_each(columns, columns + num_columns, [num_rows](gdf_column* col) {
```

Comment:
Will the compiler be smart enough to eliminate the entire for_each in a release build since the body of the lambda will be empty? If not, should probably wrap the whole thing in #ifndef NDEBUG.

Reply:
Very possible that it will just get boiled away. That said, I'd prefer not to wrap the whole thing in an ifdef because it's likely that other checks could be added in the future, or the `assert`s code be changed to `throw`s. Plus if its not optimized away, I'm not really concerned about the overhead of an empty `for_each`.